coalib/core/Bear.py

```diff
@@ -0,0 +1,502 @@
+        return inspect.getfile(cls)
+
+    @classproperty
+    def maintainers(cls):
+        """
+        :return:
+            A set containing ``MAINTAINERS`` if specified, else takes
+            ``AUTHORS`` by default.
+        """
+        return cls.AUTHORS if cls.MAINTAINERS == set() else cls.MAINTAINERS
+
+    @classproperty
+    def maintainers_emails(cls):
+        """
+        :return:
+            A set containing ``MAINTAINERS_EMAILS`` if specified, else takes
+            ``AUTHORS_EMAILS`` by default.
+        """
+        return (cls.AUTHORS_EMAILS
+                if cls.MAINTAINERS_EMAILS == set() else
+                cls.MAINTAINERS_EMAILS)
+
+    @enforce_signature
+    def __init__(self, section: Section, file_dict: dict):
+        """
+        Constructs a new bear.
+
+        :param section:
+            The section object where bear settings are contained.
+        :param file_dict:
+            The file-dictionary containing a mapping of filenames to the
+            according file contents.
+        :raises RuntimeError:
+            Raised when bear requirements are not fulfilled.
+        """
+        self.section = section
+        self.file_dict = file_dict
+
+        self._dependency_results = {}
+
+        self.setup_dependencies()
+        cp = type(self).check_prerequisites()
+        if cp is not True:
+            error_string = ('The bear ' + self.name +
+                            ' does not fulfill all requirements.')
+            if cp is not False:
+                error_string += ' ' + cp
+
+            raise RuntimeError(error_string)
+
+    def reset_dependency_results(self):
+        """
+        Resets all dependency results for this bear.
+
+        >>> section = Section('my-section')
+        >>> file_dict = {'file1.txt': ['']}
+        >>> bear = Bear(section, file_dict)
+        >>> dependency_bear = Bear(section, file_dict)
+        >>> bear.add_dependency_results(dependency_bear, [1, 2, 3])
+        >>> bear.dependency_results
+        {<class 'coalib.core.Bear.Bear'>: [1, 2, 3]}
+        >>> bear.reset_dependency_results()
+        >>> bear.dependency_results
+        {}
+        """
+        self._dependency_results = {}
+
+    def add_dependency_results(self, bear, dependency_results):
+        """
+        Adds dependency results to this instance.
+
+        This function is used by the core to add dependency results as they
+        are ready.
+
+        >>> section = Section('my-section')
+        >>> file_dict = {'file1.txt': ['']}
+        >>> bear = Bear(section, file_dict)
+        >>> bear.dependency_results
+        {}
+        >>> dependency_bear = Bear(section, file_dict)
+        >>> bear.add_dependency_results(dependency_bear, [1, 2, 3])
+        >>> bear.dependency_results
+        {<class 'coalib.core.Bear.Bear'>: [1, 2, 3]}
+
+        :param bear:
+            The bear instance to add results for.
+        :param dependency_results:
+            The results to add.
```

Comment:
Maybe mention `An iterable of results to add.` because later in line 251, doing `list(dependency_result)` will lead to an `TypeError` or unwanted results if a non-iterable or string is passed.